File: android/src/main/java/com/superwallreactnative/models/SubscriptionStatus.kt
package com.superwallreactnative.models

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReadableMap
import com.superwall.sdk.models.entitlements.SubscriptionStatus


fun SubscriptionStatus.toJson(): ReadableMap {
  val status = this
  val map = Arguments.createMap()
  map.putString("status", when (status) {
    is SubscriptionStatus.Active -> "ACTIVE"
    is SubscriptionStatus.Inactive -> "INACTIVE"
    is SubscriptionStatus.Unknown -> "UNKNOWN"
  })
  map.putArray("entitlements", Arguments.createArray().apply {
    when (status) {
      is SubscriptionStatus.Active -> {
        status.entitlements.forEach { entitlement ->
          val entitlementMap = Arguments.createMap()
          entitlementMap.putString("id", entitlement.id)
          pushMap(entitlementMap)
        }
      }
      is SubscriptionStatus.Inactive -> {}
      is SubscriptionStatus.Unknown -> {}
    }
  })
  return map
}

File: src/public/SubscriptionStatus.ts
import { Entitlement } from './Entitlement';

export type SubscriptionStatus =
  | SubscriptionStatus.Active
  | SubscriptionStatus.Inactive
  | SubscriptionStatus.Unknown;

export namespace SubscriptionStatus {
  export type Active = {
    status: `ACTIVE`;
    entitlements: Entitlement[];
  };

  export type Inactive = {
    status: `INACTIVE`;
  };

  export type Unknown = {
    status: `UNKNOWN`;
  };

  export function Active(input: Entitlement[] | string[]): Active {
    return {
      status: `ACTIVE`,
      entitlements:
        input.length === 0
          ? []
          : typeof input[0] === 'string'
          ? (input as string[]).map((id) => new Entitlement(id))
          : (input as Entitlement[]),
    };
  }

  export function Inactive(): Inactive {
    return {
      status: 'INACTIVE',
    };
  }

  export function Unknown(): Unknown {
    return {
      status: 'UNKNOWN',
    };
  }

  export function fromString(
    value: string,
    entitlements: Entitlement[]
  ): SubscriptionStatus {
    switch (value) {
      case 'ACTIVE':
        return Active(entitlements);
      case 'INACTIVE':
        return Inactive();
      case 'UNKNOWN':
      default:
        return Unknown();
    }
  }

  export function fromJson(json: any): SubscriptionStatus {
    switch (json.status) {
      case 'ACTIVE':
        return {
          status: 'ACTIVE',
          entitlements: json.entitlements.map((entitlement: any) =>
            Entitlement.fromJson(entitlement)
          ),
        };
      case 'INACTIVE':
        return {
          status: 'INACTIVE',
        };
      case 'UNKNOWN':
      default:
        return {
          status: 'UNKNOWN',
        };
    }
  }
}
